File: app/src/main/java/re/snapwi/orientation/JokePresenter.java
package re.snapwi.orientation;

import android.os.Bundle;
import dagger.Module;
import dagger.Provides;
import dagger.Subcomponent;
import de.greenrobot.event.EventBus;
import icepick.Icepick;
import icepick.Icicle;
import javax.inject.Inject;
import re.snapwi.orientation.di.PerActivity;
import re.snapwi.orientation.event.JokeEvent;
import re.snapwi.orientation.io.Joke;
import re.snapwi.orientation.ui.MainActivity;

public class JokePresenter {

  public static final String JOKE_KEY = "JokeKeyasdfasdf";

  EventBus bus;
  JokeListener listener;

  @Inject @PerActivity public JokePresenter(EventBus bus, JokeListener listener) {
    this.bus = bus;
    this.listener = listener;
  }

  @Icicle Joke joke;
  @Icicle boolean isLoadingJoke;

  public void onResume() {
    bus.registerSticky(this);
  }

  public void onPause() {
    bus.unregister(this);
  }

  public void restoreState(Bundle state) {
    Icepick.restoreInstanceState(this, state);

    if (state == null && !isLoadingJoke) {
      getRandomJoke();
    }

    if (joke != null) {
      listener.onJokeLoaded(joke);
    }
  }

  public void onEventMainThread(JokeEvent event) {
    if (event.hasError()) {
      listener.onFailed();
    } else {
      joke = event.getJoke();
      listener.onJokeLoaded(joke);
    }

    isLoadingJoke = false;
    bus.removeStickyEvent(event);
  }

  public void saveState(Bundle outstate) {
    Icepick.saveInstanceState(this, outstate);
  }

  public void getRandomJoke() {
    listener.onStartedLoading();
    isLoadingJoke = true;
    bus.post(new NorrisApiController.GetRandomJokeEvent());
  }

  public interface JokeListener {
    void onJokeLoaded(Joke joke);

    void onFailed();

    void onStartedLoading();
  }

  @PerActivity @Subcomponent(modules = JokeModule.class)
  public interface JokeComponent {

    void inject(MainActivity activity);
  }

  @Module public static class JokeModule {
    JokeListener listener;

    public JokeModule(JokeListener listener) {
      this.listener = listener;
    }

    @Provides @PerActivity public JokeListener provideListener() {
      return listener;
    }
  }
}

File: app/src/main/kotlin/re/snapwi/orientation/ui/JokePresenter.kt
package re.snapwi.orientation.ui

import android.os.Bundle
import de.greenrobot.event.EventBus
import re.snapwi.orientation.di.PerActivity
import re.snapwi.orientation.event.JokeEvent
import re.snapwi.orientation.io.Joke
import re.snapwi.orientation.rest.NorrisApiController
import javax.inject.Inject

public class JokePresenter
@Inject @PerActivity constructor(var bus: EventBus, var listener: JokePresenter.JokeListener) {

  val JOKE_KEY = "JOKE_KEY"
  val LOADING_JOKE_KEY = "LoadingJokeKey"
  var joke: Joke? = null
  var isLoadingJoke: Boolean = false

  public fun onResume(): Unit = bus.registerSticky(this)

  public fun onPause(): Unit = bus.unregister(this)

  fun saveState(outState: Bundle) {
    outState.putParcelable(JOKE_KEY, joke)
    outState.putBoolean(LOADING_JOKE_KEY, isLoadingJoke)
  }

  public fun restoreState(state: Bundle) {
    joke = state.getParcelable(JOKE_KEY)
    isLoadingJoke = state.getBoolean(LOADING_JOKE_KEY, false)

    if (state.isEmpty() && !isLoadingJoke) {
      getRandomJoke()
    }

    if (joke != null) {
      listener.onJokeLoaded(joke as Joke)
    }
  }

  public fun onEventMainThread(event: JokeEvent) {
    if (event.hasError()) {
      listener.onFailed()
    } else {
      joke = event.joke
      listener.onJokeLoaded(joke as Joke)
    }

    isLoadingJoke = false
    bus.removeStickyEvent(event)
  }

  public fun getRandomJoke() {
    listener.onStartedLoading()
    isLoadingJoke = true
    bus.post(NorrisApiController.GetRandomJokeEvent())
  }

  public interface JokeListener {
    fun onJokeLoaded(joke: Joke)

    fun onFailed()

    fun onStartedLoading()
  }
}